tests/controls/gridtest.cpp

```diff
@@ -31,32 +31,165 @@
+wxString CellCoordsToString(int row, int col)
+{
+    return wxString::Format("(%d, %c)", 1 + row, 'A' + col);
+}
+
+wxString CellSizeToString(int rows, int cols)
+{
+    return wxString::Format("%dx%d", rows, cols);
+}
+
+struct Multicell
+{
+    int row, col, rows, cols;
+
+    Multicell(int row, int col, int rows, int cols)
+        : row(row), col(col), rows(rows), cols(cols) { }
+
+    wxString Coords() const { return CellCoordsToString(row, col); }
+
+    wxString Size() const { return CellSizeToString(rows, cols); }
+
+    wxString ToString() const
+    {
+        wxString s;
+
+        if ( rows == 1 && cols == 1 )
+            s = "cell";
+        else
+        {
+            s = Size() + " ";
+            if ( rows > 1 || cols > 1 )
+                s += "multicell";
+            else
+                s += "inside cell";
+        }
+
+        return wxString::Format( "%s at %s", s, Coords() );
+    }
+};
+
+// Stores insertion/deletion info of rows or columns.
+struct EditInfo
+{
+    int pos, count, isRow;
+
+    EditInfo(int pos, int count, int isRow)
+        : pos(pos), count(count), isRow(isRow) { }
+};
+
+// Derive a new class inheriting from wxGrid, also to get access to its
 // protected GetCellAttr(). This is not pretty, but we don't have any other way
 // of testing this function.
 class TestableGrid : public wxGrid
 {
 public:
     explicit TestableGrid(wxWindow* parent)
-        : wxGrid(parent, wxID_ANY)
+        : wxGrid(parent, wxID_ANY), m_edit(0, 0, 0)
     {
     }
 
     wxGridCellAttr* CallGetCellAttr(int row, int col) const
     {
         return GetCellAttr(row, col);
     }
+
+    bool HasAttr(int row, int col,
+                 wxGridCellAttr::wxAttrKind kind = wxGridCellAttr::Cell) const
+    {
+        // Can't use GetCellAttr() here as it always returns an attr.
+        wxGridCellAttr* attr = GetTable()->GetAttr(row, col, kind);
+        if ( attr )
+            attr->DecRef();
+
+        return attr != NULL;
+    }
+
+    size_t GetCellAttrCount() const
+    {
+        // Note that only attributes in grid range can easily be checked
+        // and this function only counts those, not any outside of
+        // the grid (e.g. with invalid negative coords).
+        size_t count = 0;
+        for ( int row = 0; row < GetNumberRows(); ++row )
+            for ( int col = 0; col < GetNumberCols(); ++col )
+                count += HasAttr(row, col);
+
+        return count;
+    }
+
+    void SetMulticell(const Multicell& multi)
+    {
+        SetCellSize(multi.row, multi.col, multi.rows, multi.cols);
+    }
+
+    // Performs given insertions/deletions on either rows or columns.
+    void DoEdit(const EditInfo& edit);
+
+    // Returns annotated grid represented as a string.
+    wxString ToString() const;
+
+    // Used when drawing annotated grid to know what happens to it.
+    EditInfo m_edit;
+
+    // Grid as string before editing, with edit info annotated.
+    wxString m_sOriGridAnnotated;
```

Comment:
It's difficult to be sure, but shouldn't this be "OrigGrid", i.e. with 2 "g"s? Also, if "s" prefix means "string", I'd remove it if you're renaming this variable anyhow because we don't use Hungarian notation normally (I wouldn't mention either of these small problems on its own, but together they just justify a comment...).

Reply:
The name was trying to refer to the original grid (before edits) rendered as a string, with (to be) insertions/deletions mentioned in the side-lines.
I use `ori`(ginal) myself and forgot to change this after seeing wx had `orig` usages. That's still ambiguous though (e.g. in wx `orig[inal]Width` vs. `orig[in]ScaleX`) so maybe just something with `original` or `before`. `old` could also work but that seemed less precise.

I'll drop the `s` prefix. I use it (well, not for static strings...) or `str` sometimes only for strings in usually short-lived code when the same thing is represented by different types. E.g. `width` as integer and string, where I dislike alternatives such as `widthStr`, `widthAsStr`, or `s_width` even more. Though here I suppose it can already be more clear that it's a string from the postfix which is `Annotated` right now, I'd still like to know in case there's some preference in wx for such cases.